### PR TITLE
[Shepherd] Treat Cygwin as a separate OS

### PR DESCRIPTION
**Pull request by CodaFi**
To re-up [the discussion](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20160502/016450.html) that terminated with Chris believing that this kind of thing could only help, not hurt, I've rebased #2351.

// cc @tinysun212

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by tinysun212** [comment]
Thanks to re-up os(Cygwin). I thought that my past PR is dying.

Many POSIX-compliant systems return the operating system name with the POSIX function `uname()` and shell command `uname`. Cygwin also has those.
If we query the OS name under Cygwin environment, it returns 'Cygwin'.
```
$ uname -o
Cygwin
```
POSIX-compliant applications that run on Cygwin should recognize its OS as 'Cygwin' without any special logic.
I think that the Cygwin port of Swift compiler is a POSIX application and it can be natural to recognize the OS name to be the output of the `uname` under POSIX system. 

If this is merged, the macro definition used in Glibc port and Foundation port can be replaced in the cleaner way. swiftpm port (https://github.com/apple/swift-package-manager/pull/733) also can be reopened.

**Comment by CodaFi** [comment]
The base patch is also incomplete.  There are numerous places where we guard on `os(Windows) || CYGWIN` (or, more often than not, `!os(Windows) || CYGWIN`) that need to be changed.

**Comment by CodaFi** [comment]
@tinysun212 I've re-arranged the platform conditionals throughout the standard library and stdlib unit test to align with this patch.  Can you please run the unit tests on a Cygwin build and see if I've accidentally regressed something?

File: lib/Basic/LangOptions.cpp [hunk 1]
```diff
@@ -157,6 +158,8 @@ std::pair<bool, bool> LangOptions::setTarget(llvm::Triple triple) {
     addPlatformConditionValue(PlatformConditionKind::OS, "FreeBSD");
   else if (triple.isOSWindows())
     addPlatformConditionValue(PlatformConditionKind::OS, "Windows");
+  else if (triple.isWindowsCygwinEnvironment())
+    addPlatformConditionValue(PlatformConditionKind::OS, "Cygwin");
```
Comment (tinysun212) [inline lib/Basic/LangOptions.cpp]: On Cygwin, `triple.isOSWindows()` and `triple.isWindowsCygwinEnvironment()` are all `true`.
The condition order should be changed. After changed, `os(Windows)` will be `false` and only `os(Cygwin)` will be `true`.

File: stdlib/public/core/Policy.swift [hunk 2]
```diff
@@ -104,7 +104,7 @@ public typealias StringLiteralType = String
 // IEEE Binary64, and we need 1 bit to represent the sign.  Instead of using
 // 1025, we use the next round number -- 2048.
 public typealias _MaxBuiltinIntegerType = Builtin.Int2048
-#if (!os(Windows) || CYGWIN) && (arch(i386) || arch(x86_64))
+#if os(Cygwin) && (arch(i386) || arch(x86_64))
```
Comment (tinysun212) [inline stdlib/public/core/Policy.swift]: Just drop `CYGWIN`.
`#if !os(Windows) && (arch(i386) || arch(x86_64))`

Linux i386/x86_64, Cygwin i386/x86_64 and macOS x86_64 have `Builtin.FPIEEE80`.
Windows i386/x86_64 and other CPUs do not have.

File: stdlib/public/core/CTypes.swift [hunk 3]
```diff
@@ -43,14 +43,14 @@ public typealias CShort = Int16
 public typealias CInt = Int32
 
 /// The C 'long' type.
-#if os(Windows) && !CYGWIN && arch(x86_64)
+#if os(Windows) && !os(Cygwin) && arch(x86_64)
```
Comment (tinysun212) [inline stdlib/public/core/CTypes.swift]: Old `os(Windows) && !CYGWIN` is equivalent to new `os(Windows)`.
Just drop `&& !CYGWIN`.
Change to `#if os(Windows) && arch(x86_64)`

File: stdlib/public/core/CTypes.swift [hunk 4]
```diff
@@ -43,14 +43,14 @@ public typealias CShort = Int16
...
 public typealias CLong = Int32
 #else
 public typealias CLong = Int
 #endif
 
 /// The C 'long long' type.
-#if os(Windows) && !CYGWIN && arch(x86_64)
+#if os(Windows) && !os(Cygwin) && arch(x86_64)
```
Comment (tinysun212) [inline stdlib/public/core/CTypes.swift]: Same as above comment.

**Comment by tinysun212** [comment]
@CodaFi  you missed https://github.com/apple/swift/blob/master/stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb#L23 and https://github.com/apple/swift/blob/master/stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb#L1500.

**Comment by tinysun212** [comment]
That's all.
I cherry-picked this PR (with my comments) to my working Cygwin port and not found any regression.

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift [hunk 5]
```diff
@@ -13,13 +13,15 @@
...
+#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin)
 import Glibc
 #endif
 
 
-#if !os(Windows) || CYGWIN
-// posix_spawn is not available on Android or Windows.
+#if os(Windows)
+// posix_spawn is not available on Windows.
+#else
```
Comment (jrose-apple) [inline stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift]: Better to collapse these two conditionals, I think. There's nothing particularly special about Windows or Android.

**Review by jrose-apple — APPROVED** [review]

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Comment by tinysun212** [comment]
I'm very glad as if my old PR was passed.
Thanks @CodaFi

**Pull request merged**